### ISSUE TITLE
Feature Request: support encoding

### ISSUE THREAD

**Issue by shawalli** [ggicci/httpin#56]
This package is awesome for a service receiving requests and unmarshalling them into Go structs. But it would be even better to be able to have a single struct for clients _and_ services to use to marshal a request into an `http.Request` struct, and unmarshal a `http.Request` struct into a Go struct.

I played around with this idea myself and I'm open to contributing if it is approved as an idea.

**Comment by ggicci** [comment]
Thanks for the proposal @shawalli! Actually this is the missing feature that some other similar projects has provided, e.g. [gorilla/schema](https://github.com/gorilla/schema). It would be great for httpin to have its own to make its user experience better. Contributions are most welcome! I'm glad to see and will offer any support I can.

**Comment by shawalli** [comment]
I've started working on this, but I have a few questions I'd like your input on:

- How do you see users encoding a request? Something like `func NewRequest(method string, url string, value interface{}) (*http.Request, error)`? Or something else?
- Directives seem to be built for one-way actions (i.e. decoding). To rename it with more specificity would require a breaking change. To change The `DirectiveExecutor` interface would require a breaking change. I feel like I may just be thinking about this incorrectly...is there a better way to tackle this part?
- I played around with `func NewRequest(method string, url string, value interface{}) (*http.Request, error)`, e.g. `req, err := NewRequest("GET", "/users/{user_id}/resources/{resource_id}, &MyRequestStruct)`, where the function would render a final `http.Request.URL` using the struct's directives, set headers, and encode the body. How do you feel about that direction?

**Comment by ggicci** [comment]
> How do you see users encoding a request? Something like func NewRequest(method string, url string, value interface{}) (*http.Request, error)? Or something else?

It looks good to me!

> Directives seem to be built for one-way actions (i.e. decoding). To rename it with more specificity would require a breaking change. To change The DirectiveExecutor interface would require a breaking change. I feel like I may just be thinking about this incorrectly...is there a better way to tackle this part?

I would recommend adding another separate interface to archieve the functionality that populating an http request with the field values of the struct instance. For example,

```go
type DirectiveRequestPopulater interface {
    Populate(ctx RequestPopulaterContext, req *http.Request, fieldValue interface{}) error
}
```

I believe [resolve](https://github.com/ggicci/httpin/blob/f9557f5f710e24a6f89c95f2d126c6dace16199d/httpin.go#L106) is a good start point to look into. Where httpin iterates each struct field and runs the directives one by one. But here we are expecting to run the directives in a "reverse" way. Originally it reads from request and sets struct field values. Now, it populates the corresponding request properties with struct field values. Since the [tree has been built](https://github.com/ggicci/httpin/blob/f9557f5f710e24a6f89c95f2d126c6dace16199d/httpin.go#LL115C19-L115C19), you can reuse it and feel free to make some tweaks to fit your needs.

We need to implement the corresponding built-in populaters (reverse directive executors), i.e. "form", "query", etc. And don't forget to **register** these populaters to httpin, too.

When we call `NewRequest()`, we should run like the decoding process, but in the reverse logic.

> I played around with func NewRequest(method string, url string, value interface{}) (*http.Request, error), e.g. req, err := NewRequest("GET", "/users/{user_id}/resources/{resource_id}, &MyRequestStruct), where the function would render a final http.Request.URL using the struct's directives, set headers, and encode the body. How do you feel about that direction?

Yes, it looks awesome!

**Comment by ggicci** [comment]
Encoding feature is implemented since release [v0.15.0](../releases/v0.15.0).
